### PR TITLE
Proposal to check code style a bit earlier

### PR DESCRIPTION
**Pull request by AndreasAntener**
I've separated the list of included modules to check code style so everyone can include it in their git pre-commit hook and run it for changed files only before the commit. This should get rid of these unnecessary "code style" commits at the end of a pull request.

Had to remove `-print0` from find, which should be fine as long as we don't have spaces in the paths.. Maybe a linux guru can suggest a way to keep `-print0` (and test this :)).

Once this is good I can followup with a quick doc in the devguide.

How-to, in your repositories:
```
mv .git/hooks/pre-commit.sample .git/hooks/pre-commit
chmod 755 .git/hooks/pre-commit
```

Then edit `.git/hooks/pre-commit` and replace the white-space check (yes, it also does that 👍 ) at the end to:
```
# If there are whitespace errors, print the offending file names and fail.
git diff-index --check --cached $against --
if [ $? -ne 0 ]
then
        exit 1
fi

# Check for code style, only in changed files
for i in `git diff --cached --name-only --diff-filter=ACM`
do
        ./Tools/files_to_check_code_style.sh $i | xargs -n 1 -P 8 -I % ./Tools/check_code_style.sh %
        if [ $? -ne 0 ]
        then
                exit 1
        fi
done
```

(see my complete pre-commit file: https://gist.github.com/AndreasAntener/0f9cc67cd4d4eb321f935297d9c268bd)

### CODE REVIEW

**Comment by dagar** [comment]
I guess I should have done a better job publicizing this...

https://github.com/PX4/Firmware/blob/master/Tools/pre-commit

**Comment by dagar** [comment]
Yours looks better. If we wanted to get fancy we could have `make check_format` prompt to install the commit hook (copy + perms).

**Comment by AndreasAntener** [comment]
Sorry I was also blind it seems...
I like your fancy idea :). A good start could be to let `check_code_style_all.sh` do that?

**Comment by dogmaphobic** [comment]
@dagar We should probably do the same for QGC. It's a pain going back and forth between QGC and Firmware as they differ in code style quite a bit (especially tab versus spaces).

**Comment by dagar** [comment]
@dogmaphobic I'm all for it if you and @DonLakeFlyer can tolerate the massive diff. Then I'd just use the same pieces from PX4 astyle config, make check_format, style enforced by travis-ci, etc.

@AndreasAntener these scripts keep growing, but if you put it in the lowest level (./Tools/fix_code_style.sh) it'll be the most annoying.

**Comment by dogmaphobic** [comment]
> can tolerate the massive diff

I don't have any issues with it if @DonLakeFlyer doesn't as well.

**Comment by DonLakeFlyer** [comment]
@dagar Can you open an issue on the QGC side and we can continue discussion there?

**Comment by LorenzMeier** [comment]
@AndreasAntener Nice! given that this really would benefit from a pre-commit hook, is there a chance to have the prompt to be added?

**Comment by AndreasAntener** [comment]
Looking into it, need to find a way so it's not tooo annoying.

**Comment by AndreasAntener** [comment]
@dagar @LorenzMeier added prompt on highest level (further down it gets swallowed).

**Comment by AndreasAntener** [comment]
Oh, this probably breaks CI, need to exclude the check there

**Comment by dagar** [comment]
Thanks @AndreasAntener, looks good. One last request that I'm happy to do myself later if you aren't interested. Can we merge these into one script? Having multiple dependent scripts like this is getting hard to follow.

    Tools/check_code_style.sh
    Tools/check_code_style_all.sh
    Tools/files_to_check_code_style.sh

You could even pull in `./Tools/fix_code_style.sh`.
Then the behavior could be...
   - `./Tools/check_code_style.sh`: check all files and only print diff (make check_format)
   - `./Tools/check_code_style.sh $FILE`: check only specified file
   - `./Tools/check_code_style.sh --fix` print diff and fix in place (make format)

**Comment by AndreasAntener** [comment]
Sure if you'd like to do that I have no objections (leaving this branch to you). But the reason to extract the file list was the hook. So for the pre-commit hook ` ./Tools/check_code_style.sh $FILE` should return successfully if the file is not in the list of to be checked files.

**Comment by AndreasAntener** [comment]
@dagar already in my last commit Travis didn't seem to like the $CI variable but it should be there according to the docs: https://docs.travis-ci.com/user/environment-variables/. Not sure why, maybe we have to try something else.

**Comment by AndreasAntener** [comment]
@dagar do you have an idea about how to exclude Travis? If we have that I guess we can merge it.

**Comment by dagar** [comment]
@AndreasAntener we need to passthru CI=true to docker. I can finish it tomorrow.
travis-ci, semaphoreci, and circleci

I hope you don't mind I made some changes. I instinctively ignore large blocks of text so I tried to keep it terse.

**Comment by AndreasAntener** [comment]
Ah that's why...
I was wondering if we should leave the command in the print so people see what it does before they confirm.

**Comment by dagar** [comment]
@AndreasAntener should be working now, please make sure I didn't break anything. Feel free to squash my commits into yours and merge, I actually prefer the simple history.

**Comment by AndreasAntener** [comment]
Great, thanks! I'll squash quickly

**Comment by AndreasAntener** [comment]
Rebased and merged to master
